statistics/feedback.go

```diff
@@ -333,6 +293,68 @@ func (b *BucketFeedback) splitBucket(newBktNum int, totalCount float64, count fl
 	return bkts
 }
 
+func getFraction4PK(minValue, maxValue, lower, upper *types.Datum) (float64, float64) {
+	if minValue.Kind() == types.KindInt64 {
+		l, r := float64(minValue.GetInt64()), float64(maxValue.GetInt64())
+		return calcFraction(l, r, float64(lower.GetInt64())), calcFraction(l, r, float64(upper.GetInt64()))
+	}
+	l, r := float64(minValue.GetUint64()), float64(maxValue.GetUint64())
+	return calcFraction(l, r, float64(lower.GetUint64())), calcFraction(l, r, float64(upper.GetUint64()))
+}
+
+func getFraction4Index(minValue, maxValue, lower, upper *types.Datum, prefixLen int) (float64, float64) {
+	l, r := convertBytesToScalar(minValue.GetBytes()[prefixLen:]), convertBytesToScalar(maxValue.GetBytes()[prefixLen:])
+	return calcFraction(l, r, convertBytesToScalar(lower.GetBytes()[prefixLen:])),
+		calcFraction(l, r, convertBytesToScalar(upper.GetBytes()[prefixLen:]))
+}
+
+// getOverlapFraction gets the overlap fraction of feedback and bucket range. In order to get the bucket count, it also
+// returns the ratio between feedback fraction and bucket fraction.
+func getOverlapFraction(fb feedback, bkt bucket) (float64, float64) {
+	datums := make([]types.Datum, 0, 4)
+	datums = append(datums, *fb.lower, *fb.upper)
+	datums = append(datums, *bkt.lower, *bkt.upper)
+	err := types.SortDatums(nil, datums)
+	if err != nil {
+		return 0, 0
+	}
+	var fbLower, fbUpper, bktLower, bktUpper float64
+	minValue, maxValue := &datums[0], &datums[3]
+	if datums[0].Kind() == types.KindBytes {
+		prefixLen := commonPrefixLength(minValue.GetBytes(), maxValue.GetBytes())
+		fbLower, fbUpper = getFraction4Index(minValue, maxValue, fb.lower, fb.upper, prefixLen)
+		bktLower, bktUpper = getFraction4Index(minValue, maxValue, bkt.lower, bkt.upper, prefixLen)
+	} else {
+		fbLower, fbUpper = getFraction4PK(minValue, maxValue, fb.lower, fb.upper)
+		bktLower, bktUpper = getFraction4PK(minValue, maxValue, bkt.lower, bkt.upper)
+	}
+	ratio := (bktUpper - bktLower) / (fbUpper - fbLower)
+	// full overlap
+	if fbLower <= bktLower && bktUpper <= fbUpper {
+		return bktUpper - bktLower, ratio
+	}
+	if bktLower <= fbLower && fbUpper <= bktUpper {
+		return fbUpper - fbLower, ratio
+	}
+	// partial overlap
+	overlap := math.Min(bktUpper-fbLower, fbUpper-bktLower)
+	return overlap, ratio
+}
+
+func (b *BucketFeedback) bucketCount(bkt bucket, defaultCount float64) float64 {
+	bestFraction := minBucketFraction
+	count := defaultCount
+	for _, fb := range b.feedback {
+		fraction, ratio := getOverlapFraction(fb, bkt)
+		// choose the max overlap fraction
+		if fraction > bestFraction {
+			bestFraction = fraction
+			count = float64(fb.count) * ratio
+		}
+	}
+	return count
+}
+
 // Get the split count for the histogram.
 func getSplitCount(count, remainBuckets int) int {
 	remainBuckets = mathutil.Max(remainBuckets, 10)
```

Comment:
better use another variable like `splitCount` rather than modify the input argument.